File: yggdrasil/examples/timesync2/src/timesync.f90
program main
  use fygg

  ! Declare resulting variables and create buffer for received message
  logical :: timestep_calc
  logical :: ret = .true.
  type(yggcomm) :: timesync, out
  character(len=32) :: arg
  character(len=0), dimension(0) :: dtype_keys
  type(yggdtype), dimension(0) :: dtype_vals
  type(yggdtype) :: out_dtype
  real(kind=8) :: t_step, t_start, t_end, t
  character(len=32) :: t_units, model
  character(len=20), dimension(:), pointer :: keys
  type(ygggeneric) :: state_send, state_recv, msg
  real(kind=8), pointer :: x
  integer :: i

  call get_command_argument(1, arg)
  read(arg, *) t_step
  call get_command_argument(2, arg)
  read(arg, *) t_units
  call get_command_argument(3, arg)
  read(arg, *) model
  write (*, '("Hello from Fortran timesync: timestep ",F10.5," ",A3,&
       &" (model = ",A,")")') t_step, trim(t_units), trim(model)
  t_start = 0.0
  t_end = 5.0
  if (t_units.eq."hr") then
     t_end = 24.0 * t_end
  end if
  state_send = init_generic_map()
  state_recv = init_generic_map()
  ret = timestep_calc(t_start, t_units, state_send, model)
  if (.not.ret) then
     write (*, '("timesync(Fortran): Error in initial timestep &
          &calculation.")')
  end if

  ! Set up connections matching yaml
  ! Timestep synchronization connection will be 'statesync'
  timesync = ygg_timesync("statesync", t_units)
  out_dtype = create_dtype_json_object(0, dtype_keys, dtype_vals, .true.)
  out = ygg_output_type("output", out_dtype)

  ! Initialize state and synchronize with other models
  t = t_start
  ret = ygg_rpc_call(timesync, [yggarg(t), yggarg(state_send)], &
       yggarg(state_recv))
  if (.not.ret) then
     write (*, '("timesync(Fortran): Initial sync failed.")')
     stop 1
  end if
  write (*, '("timesync(Fortran): t = ",F5.1," ",A3)', advance="no") &
       t, adjustl(t_units)
  call generic_map_get_keys(state_recv, keys)
  do i = 1, size(keys)
     call generic_map_get(state_recv, trim(keys(i)), x)
     write (*, '(SP,", ",A," = ",F5.2)', advance="no") &
          trim(keys(i)), x
  end do
  write (*, '("")')

  ! Send initial state to output
  msg = copy_generic(state_recv)
  call generic_map_set(msg, "time", t, t_units)
  ret = ygg_send_var(out, yggarg(msg))
  if (.not.ret) then
     write (*, '("timesync(Fortran): Failed to send initial output &
          &for t=",F10.5,".")') t
     stop 1
  end if
  call free_generic(msg)

  ! Iterate until end
  do while (t.lt.t_end)

     ! Perform calculations to update the state
     t = t + t_step
     ret = timestep_calc(t, t_units, state_send, model)
     if (.not.ret) then
        write (*, '("timesync(Fortran): Error in timestep &
             &calculation for t = ",F10.5,".")') t
        stop 1
     end if

     ! Synchronize the state
     ret = ygg_rpc_call(timesync, [yggarg(t), yggarg(state_send)], &
          yggarg(state_recv))
     if (.not.ret) then
        write (*, '("timesync(Fortran): sync failed for t=",F10.5,&
             &".")') t
        stop 1
     end if
     write (*, '("timesync(Fortran): t = ",F5.1," ",A3)', advance="no") &
          t, adjustl(t_units)
     call generic_map_get_keys(state_recv, keys)
     do i = 1, size(keys)
        call generic_map_get(state_recv, keys(i), x)
        write (*, '(SP,", ",A," = ",F5.2)', advance="no") &
             trim(keys(i)), x
     end do
     write (*, '("")')

     ! Send output
     msg = copy_generic(state_recv)
     call generic_map_set(msg, "time", t, t_units)
     ret = ygg_send_var(out, yggarg(msg))
     if (.not.ret) then
        write (*, '("timesync(Fortran): Failed to send output for &
             &t=",F10.5,".")') t
        stop 1
     end if
     call free_generic(msg)

  end do

  write (*, '("Goodbye from Fortran timesync")')
  call free_generic(state_send)
  call free_generic(state_recv)

end program main


function timestep_calc(t, t_units, state, model) result (ret)
  use fygg
  implicit none
  real(kind=8) :: t
  character(len=*) :: t_units
  type(ygggeneric) :: state
  character(len=*) :: model
  logical :: ret
  real(kind=8) :: x_period = 10.0  ! Days
  real(kind=8) :: y_period = 5.0   ! Days
  real(kind=8) :: z_period = 20.0  ! Days
  real(kind=8) :: o_period = 2.5   ! Days
  ret = .true.
  if (t_units.eq."day") then
     ! No conversion necessary
  else if (t_units.eq."hr") then
     x_period = x_period * 24.0
     y_period = y_period * 24.0
     z_period = z_period * 24.0
     o_period = o_period * 24.0
  else
     write (*, '("timestep_calc: Unsupported unit ''",A,"''")') t_units
     ret = .false.
  end if
  if (ret) then
     if (model.eq."A") then
        call generic_map_set(state, "x", &
             sin(2.0 * PI_8 * t / x_period))
        call generic_map_set(state, "y", &
             cos(2.0 * PI_8 * t / y_period))
        call generic_map_set(state, "z1", &
             -cos(2.0 * PI_8 * t / z_period))
        call generic_map_set(state, "z2", &
             -cos(2.0 * PI_8 * t / z_period))
        call generic_map_set(state, "a", &
             sin(2.0 * PI_8 * t / o_period))
     else
        call generic_map_set(state, "xvar", &
             sin(2.0 * PI_8 * t / x_period))
        call generic_map_set(state, "yvar", &
             cos(2.0 * PI_8 * t / y_period))
        call generic_map_set(state, "z", &
             -2.0 * cos(2.0 * PI_8 * t / z_period))
        call generic_map_set(state, "b", &
             cos(2.0 * PI_8 * t / o_period))
     end if
  end if
end function timestep_calc

File: yggdrasil/examples/timesync2/src/timesync.cpp
#define _USE_MATH_DEFINES  // Required to use M_PI with MSVC
#include <math.h>
#include <stdio.h>
#include "YggInterface.hpp"


void timestep_calc(rapidjson::units::Quantity<double>& t,
		   rapidjson::Document& state, std::string model) {
  rapidjson::units::Quantity<double> x_period(10.0, "days");
  rapidjson::units::Quantity<double> y_period(5.0, "days");
  rapidjson::units::Quantity<double> z_period(10.0, "days");
  rapidjson::units::Quantity<double> o_period(2.5, "days");
#define SET_(key, val)				\
  if (!state.HasMember(key))			\
    state.AddMember(key, rapidjson::Value(val).Move(),	\
		    state.GetAllocator());		\
  else							\
    state[key].SetDouble(val)
  if (model == "A") {
    SET_("x", sin(2.0 * M_PI * (t / x_period).value()));
    SET_("y", cos(2.0 * M_PI * (t / y_period).value()));
    SET_("z1", -cos(2.0 * M_PI * (t / z_period).value()));
    SET_("z2", -cos(2.0 * M_PI * (t / z_period).value()));
    SET_("a", sin(2.0 * M_PI * (t / o_period).value()));
  } else {
    SET_("xvar", sin(2.0 * M_PI * (t / x_period).value()) / 2.0);
    SET_("yvar", cos(2.0 * M_PI * (t / y_period).value()));
    SET_("z", -2.0 * cos(2.0 * M_PI * (t / z_period).value()));
    SET_("b", cos(2.0 * M_PI * (t / o_period).value()));
  }
}


int main(int argc, char *argv[]) {

  char* t_units = argv[2];
  rapidjson::units::Quantity<double> t_step(atof(argv[1]), t_units);
  std::cout << "Hello from C++ timesync: timestep " << t_step << std::endl;
  rapidjson::units::Quantity<double> t_start(0.0, t_units);
  rapidjson::units::Quantity<double> t_end(5.0, "days");
  
  std::string model(argv[3]);
  int ret;
  rapidjson::Document state_send(rapidjson::kObjectType);
  rapidjson::Document state_recv(rapidjson::kObjectType);
  timestep_calc(t_start, state_send, model);

  // Set up connections matching yaml
  // Timestep synchronization connection will be 'statesync'
  YggTimesync timesync("statesync", t_units);
  dtype_t* out_dtype = create_dtype_json_object(0, NULL, NULL, true);
  YggOutput out("output", out_dtype);

  // Initialize state and synchronize with other models
  rapidjson::units::Quantity<double> t = t_start;
  ret = timesync.call(3, t.value(), &state_send, &state_recv);
  if (ret < 0) {
    std::cerr << "timesync(C++): Initial sync failed." << std::endl;
    return -1;
  }
  std::cout << "timesync(C++): t = " << t;
  for (rapidjson::Value::MemberIterator it = state_recv.MemberBegin();
       it != state_recv.MemberEnd(); it++)
    std::cout << ", " << it->name.GetString() <<
      " = " << it->value.GetDouble();
  std::cout << std::endl;

  // Send initial state to output
  rapidjson::Document msg;
  msg.CopyFrom(state_recv, msg.GetAllocator());
  msg.AddMember("time", rapidjson::Value(t).Move(), msg.GetAllocator());
  ret = out.send(1, &msg);
  if (ret < 0) {
    std::cerr << "timesync(C++): Failed to send initial output for t=" <<
      t << std::endl;
    return -1;
  }

  // Iterate until end
  while (t < t_end) {

    // Perform calculations to update the state
    t = t + t_step;
    timestep_calc(t, state_send, model);

    // Synchronize the state
    ret = timesync.call(3, t.value(), &state_send, &state_recv);
    if (ret < 0) {
      std::cerr << "timesync(C++): sync for t=" << t << " failed" << std::endl;
      return -1;
    }
    std::cout << "timesync(C++): t = " << t;
    for (rapidjson::Value::MemberIterator it = state_recv.MemberBegin();
	 it != state_recv.MemberEnd(); it++)
      std::cout << ", " << it->name.GetString() <<
	" = " << it->value.GetDouble();
    std::cout << std::endl;

    // Send output
    msg.CopyFrom(state_recv, msg.GetAllocator());
    msg.AddMember("time", rapidjson::Value(t).Move(), msg.GetAllocator());
    ret = out.send(1, &msg);
    if (ret < 0) {
      std::cerr << "timesync(C++): Failed to send output for t=" << t << std::endl;
      return -1;
    }
  }

  std::cout << "Goodbye from C++ timesync" << std::endl;
  return 0;
    
}
